### PR TITLE
feat(lessons-cc-hook): add policy manifest loader and classification for Stage 1 shadow logging

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Implements **Stage 1 shadow logging** for the lesson-validated-core policy in the
Claude Code hook (`scripts/claude-code-hooks/match-lessons.py`). Mirrors the
work in [gptme/gptme#3455](https://github.com/gptme/gptme/pull/3455) (merged)
on the gptme side so the two runtimes produce identical classification for
the same lesson set.

This is a **shadow** change — it does **not** alter dropout behavior, matching
rules, or epsilon. It only enriches the dropout log with `policy_class` and
`policy_version` so Stage 2 (canary rollout) can be validated against real
cross-runtime data.

## Changes

- `_load_policy_manifest()`: load `state/lesson-policy/manifest.yaml` (cached,
  fail-safe to a default empty manifest on parse error or missing file).
- `_classify_lesson(lesson_path)`: classify a lesson as `validated_core` /
  `exempt` / `holdout` / `unknown` against the loaded manifest.
- `_log_dropout()`: enrich each withheld lesson record with `policy_class`
  and `policy_version`; add a top-level `policy_version` field.
- Removes the now-superseded `_log_lesson_events` function (efficacy events
  were folded into the dropout path by gptme-contrib#1371; keeping the
  duplicate would double-log).

## Manifest format

```yaml
version: 1
updated_at: "<ISO>"
validated_core: []        # populated in Stage 2+
exempt: []                # populated in Stage 2+
holdout_population: [...] # full lesson inventory (Stage 0)
```

## Verification

- All 63 tests in `tests/test_cc_match_lessons.py` pass locally.
- `_classify_lesson()` correctly returns `('holdout', 1)` for lessons in the
  `holdout_population` list and `('unknown', 1)` for unmanifested paths.
- Manual reproduction:

  ```
  $ PYTHONPATH=scripts/claude-code-hooks python3 -c "
  from match_lessons import _classify_lesson
  print(_classify_lesson('lessons/analysis/isolate-infra-rollouts-from-behavioral-signals.md'))
  # -> ('holdout', 1)
  print(_classify_lesson('lessons/nonexistent/foo.md'))
  # -> ('unknown', 1)
  "
  ```

## Out of scope (Stage 1 explicit)

- Changing dropout probability (still 20% globally).
- Populating `validated_core` / `exempt` lists (Stage 2).
- Changing lesson matching rules.

## Related

- **Policy design**: `ErikBjare/bob` `knowledge/technical-designs/lesson-validated-core-policy.md`
- **gptme side (merged)**: gptme/gptme#3455
- **Predecessor (merged)**: gptme/gptme-contrib#1371 (BM25 z-gate; the rebase
  lands cleanly on top of it).
- **Task**: `ErikBjare/bob` task `lesson-phase2-stage1-shadow`.
- **Stage 0 (manifest inventory)**: `lesson-phase2-stage0-rollout`.

### CODE REVIEW

File: scripts/claude-code-hooks/match-lessons.py [hunk 1]
```diff
@@ -205,6 +205,106 @@ def _trajectory_log_dir() -> Path:
...
+        - "validated_core": high-ROI, recommended for all sessions
+        - "exempt": safety/retention policy, exempt from dropout
+        - "holdout": under evaluation (default)
+        - "unknown": lesson not in manifest (created after manifest timestamp)
+    """
+    manifest = _load_policy_manifest()
+    policy_version = manifest.get("version", 1)
+
+    # Normalize lesson_path to the key format used in manifest (no leading/trailing slashes)
+    key = lesson_path.strip("/").replace("lessons/", "", 1).replace(".md", "")
```
Comment (TimeToBuildBob) [inline scripts/claude-code-hooks/match-lessons.py]: ❌ **P1** — The path normalization in `_classify_lesson` assumes `lesson_path` is relative to the workspace root and starts with `lessons/`, but actual lesson paths from `scan_lessons` are absolute (e.g. `/home/user/project/lessons/patterns/persistent-learning.md`). After `strip("/")`, the leading slash is removed, then `replace("lessons/", "", 1)` removes the first `lessons/` occurrence which appears after the workspace prefix, leaving `home/user/project/patterns/persistent-learning` as the key. Manifest keys are expected to be like `patterns/persistent-learning`, so the key never matches and every withheld lesson is classified as `unknown`. This makes the Stage 1 shadow logging useless for cross-runtime parity because no lesson is ever correctly tagged with its policy class. The PR verification passes a relative path, which works, but the hook passes absolute paths.

File: scripts/claude-code-hooks/match-lessons.py [hunk 2]
```diff
@@ -205,6 +205,106 @@ def _trajectory_log_dir() -> Path:
...
+
+    manifest_path = _policy_manifest_path()
+    if not manifest_path.exists():
+        # Manifest not yet created (Stage 0 not run) — default all to holdout
+        _policy_manifest_cache = {
+            "version": 1,
+            "updated_at": "",
+            "validated_core": [],
+            "exempt": [],
+            "holdout_population": [],
```
Comment (TimeToBuildBob) [inline scripts/claude-code-hooks/match-lessons.py]: ⚠️ **P2** — When the manifest file is missing or fails to load, `_load_policy_manifest` returns a default dict with `holdout_population: []`. The docstring promises 'a safe default (all lessons in holdout)', but with an empty list `_classify_lesson` will produce 'unknown' for every lesson, not 'holdout'. This means the shadow log will tag all lessons as `unknown` in the absence of a manifest, which may be mistaken for a broken manifest by downstream analysis. If the intent is to fall back to holdout, the default should treat all lessons as holdout, e.g. by making `_classify_lesson` return `holdout` when the manifest is missing, or by having a sentinel in holdout_population.

**Comment by TimeToBuildBob** [comment]
## 🤖 AI code review

Adds a lesson-policy manifest loader (_load_policy_manifest) and a lesson-path classifier (_classify_lesson) that returns validated_core, exempt, holdout, or unknown based on the manifest. Enriches dropout log records in _log_dropout with per-lesson policy_class/policy_version and a top-level policy_version. Adds tests covering path normalization and manifest edge cases.

<sub>This describes the diff as I read it — not a verdict on it. If it does not match the change, treat the findings below as suspect.</sub>

### Confidence Score: 5/5 — No findings on this head

Nothing outstanding on this head — no findings on this pass.

✅ **No findings.** The diff looks correct to me on this pass.

<details><summary><b>Files changed (2)</b> — the diff as I read it</summary>

- `scripts/claude-code-hooks/match-lessons.py` — Adds _load_policy_manifest with caching, _classify_lesson for policy classification, and enriches _log_dropout records with policy fields.
- `tests/test_cc_match_lessons.py` — Adds tests for _classify_lesson covering absolute, relative, dot-relative paths, missing manifest, non-dict YAML, and missing holdout_population key.

</details>

<details><summary>Previous review passes</summary>

| commit | score | findings | engine | when |
| --- | --- | --- | --- | --- |
| `e27a99933db5` | 3/5 | 2 | llm | 2026-08-17 23:48 UTC |
| `464020043bde` | 4/5 | 3 | llm | 2026-08-18 17:01 UTC |
| `78dc6ad97a06` | 3/5 | 1 | llm | 2026-08-18 17:32 UTC |

</details>

<sub>Reviewed `f0481c98c98e` · `openrouter/deepseek/deepseek-v4-flash-0731` · `llm` engine · 192s · reports findings; merge decisions belong to the gate · [bob#1122](https://github.com/ErikBjare/bob/issues/1122)</sub>

<sub>Maintainers: comment `@TimeToBuildBob review` on its own line for a fresh review, or `@TimeToBuildBob fix` to have a worker act on the findings above. Each fires once per comment — the 👀 reaction is the receipt.</sub>

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: scripts/claude-code-hooks/match-lessons.py (hunk 1 above)
Comment (TimeToBuildBob) [reply scripts/claude-code-hooks/match-lessons.py]: Fixed in 464020043bde —  now splits on `/lessons/` to handle absolute paths: if `/lessons/` is in the path, takes everything after the last `/lessons/` occurrence; falls back to stripping a `lessons/` prefix for relative paths. This produces the correct `patterns/persistent-learning` key regardless of how the path arrives.

File: scripts/claude-code-hooks/match-lessons.py (hunk 2 above)
Comment (TimeToBuildBob) [reply scripts/claude-code-hooks/match-lessons.py]: Fixed in 464020043bde — the default manifest (missing file or load failure) now uses `holdout_population: ["*"]` and _classify_lesson checks for `"*"` in the holdout list, returning `"holdout"` for every lesson. No lesson will be tagged `unknown` due to a missing manifest.

File: scripts/claude-code-hooks/match-lessons.py [hunk 3]
```diff
@@ -205,6 +205,115 @@ def _trajectory_log_dir() -> Path:
...
+    else:
+        key = path_str
+    key = key.replace(".md", "").strip("/")
+
+    for category in ["validated_core", "exempt"]:
+        if key in manifest.get(category, []):
+            return category, policy_version
+
+    # "*" sentinel means all lessons fall into holdout (manifest missing/unavailable)
+    holdout_pop = manifest.get("holdout_population", ["*"])
```
Comment (TimeToBuildBob) [inline scripts/claude-code-hooks/match-lessons.py]: ⚠️ **P2** — When the manifest file exists and parses as a dict but lacks the `holdout_population` key, the default `["*"]` is used, causing every lesson not in validated_core/exempt to be classified as "holdout". The docstring of `_load_policy_manifest` states the `*` sentinel is the safe default for a missing or unparseable manifest; for a present-but-incomplete manifest the intended behavior is to leave lessons as "unknown" (the function's own docstring says unknown is for lessons not in the manifest). This over-broad default biases the Stage 1 shadow-logging dataset: newly added lessons that never appear in the manifest get recorded as holdout instead of unknown, which will skew the Stage 2 validation comparison.

**Comment by TimeToBuildBob** [comment]
Fixed all three P2 findings from the AI reviewer (commit `78dc6ad9`):

**Finding 1** — `_load_policy_manifest` non-dict YAML: replaced `yaml.safe_load(f) or {}` with a temp-var + conditional (`_loaded if isinstance(_loaded, dict) else {}`), which mypy can narrow to `dict` and which correctly catches any non-dict YAML value (lists, scalars) before caching.

**Finding 2** — `holdout_population` default for present manifests: changed `.get("holdout_population", ["*"])` to `.get("holdout_population", [])` in `_classify_lesson`. The `"*"` sentinel is only correct for the missing-file / parse-error branches (which still supply it explicitly). A present manifest that simply omits the key should yield `"unknown"` for unlisted lessons, not `"holdout"`.

**Finding 3** — dot-relative path normalization: `./lessons/X/Y.md` paths fell to the else branch (neither containing `/lessons/` nor starting with `lessons/`). Added `lstrip("./")` on the stripped path before the `lessons/` prefix check.

Three regression tests added, one per finding. 69 tests pass; all pre-commit hooks green.
